pandas/src/util.pxd

```diff
@@ -59,6 +59,10 @@ cdef inline object get_value_at(ndarray arr, object loc):
 cdef inline set_value_at(ndarray arr, object loc, object value):
     cdef:
         Py_ssize_t i, sz
+
+    if not cnp.PyArray_ISWRITEABLE(arr):
```

Comment:
I might put this at a higher level with a single check per-array (and not on an individual set, even though that is technically more correct).